azure-iot-device/azure/iot/device/provisioning/pipeline/pipeline_stages_provisioning.py

```diff
@@ -48,3 +63,331 @@ def _execute_op(self, op):
+                        if query_op.retry_after is not None
+                        else constant.DEFAULT_POLLING_INTERVAL
+                    )
+                    decoded_response = self._decode_response(query_op)
+                    operation_id = self._get_operation_id(decoded_response)
+                    registration_status = self._get_registration_status(decoded_response)
+
+                    # retry after or assigning scenario
+                    if success_status_code >= 429 or registration_status == "assigning":
+
+                        self_weakref = weakref.ref(self)
+
+                        @pipeline_thread.invoke_on_pipeline_thread_nowait
+                        def do_polling():
+                            this = self_weakref()
+                            logger.info(
+                                "{stage_name}({op_name}): retrying".format(
+                                    stage_name=this.name, op_name=op.name
+                                )
+                            )
+                            op.polling_timer.cancel()
+                            op.completed = False
+                            this._execute_op(op)
+
+                        logger.warning(
+                            "{stage_name}({op_name}): Op needs retry with interval {interval} because of {error}. Setting timer.".format(
+                                stage_name=self.name,
+                                op_name=op.name,
+                                interval=polling_interval,
+                                error=error,
+                            )
+                        )
+
+                        op.polling_timer = Timer(polling_interval, do_polling)
+                        op.polling_timer.start()
+
+                    # Service success scenario
+                    else:
+                        registration_status = self._get_registration_status(decoded_response)
+                        if registration_status == "assigned" or registration_status == "failed":
+                            # process complete response here
+                            complete_registration_result = self._form_complete_result(
+                                operation_id=operation_id,
+                                decoded_response=decoded_response,
+                                status=registration_status,
+                            )
+                            op.registration_result = complete_registration_result
+                        else:
+                            error = exceptions.ServiceError(
+                                "Query Status Operation encountered an invalid registration status {status} with a status code of {status_code}".format(
+                                    status=registration_status, status_code=success_status_code
+                                )
+                            )
+
+                        self.complete_op(op, error=error)
```

Comment:
>complete_op [](start = 29, length = 11)

same comment as above.  If registration_status== "failed", error should be set on this completion

Reply:
I have set an additional error as well now. The polling machine implementation actually did not have this error. Hence I skipped it. I also do not know (and i have asked the DPS team) what is the nature of response from the service when the status is `failed`

---
In reply to: [345918256](https://github.com/Azure/azure-iot-sdk-python/pull/391#discussion_r345918256) [](ancestors = 345918256)